### DISCUSSION TITLE
Include textured models

### DISCUSSION THREAD

**Discussion by marcoma9023** [pelicanmapping/osgearth#2304] [Support Q&A] [answered]
Good morning,

I need to integrate 3D models of any format into my scene, I have no limitations, for example OBJ, STL, 3ds.

Previously, with OSGEarth 2.x I had no problems and each model was created with materials with color and/or texture applied.

I know that under GL3 everything is handled with shaders.

I need to understand:

1) What format for a 3D model is usually used with OSGEarth 3.x?
2) If the OBJ goes well, do you have a shader ready to handle materials?

Thanks for your attention

**Comment by gwaldron** [comment]
If you load your models through OSGEarth it will typically run them through a shader generator.
If you load them with OSG directly, you will need to do it yourself either by using the API (`osgEarth::ShaderGenerator`) or by using the `shadergen` pseudo-loader, e.g.
```c++
osgDB::readRefNodeFile("model.obj.osgearth_shadergen");
```
Hope this helps.

**Reply by marcoma9023** [reply]
I usually load the model with osg through the osgDB::readNodeFile() call

but, in case my OBJ contains materials, I have warnings that they don't know how to handle materials and the model color is black.

I tried doing as you suggested, then calling the ShaderGenerator, but the model is black.

What am I missing?

**Reply by devil-darkside** [reply]
I have the same problem... the ShaderGenerator is not working... use the ".osgearth_shadergen" to have your model fully textured.

**Comment by gwaldron** [comment]
Hey sorry for the late reply. Can you provide an example model that we can use to test?

**Reply by devil-darkside** [reply]
Use the cow.osg from "OpenSceneGraph-Data" and the code in #2328

**Comment by gwaldron** [comment, marked as answer]
Resolution in #2328.

**Reply by XiaobaiSW** [reply]
Hello, I am currently encountering another issue. I am using osg3.4 and osgearth2.7, osg:: Node * pNode=osgDB:: readRefNodeFile ("cow. osg") or osg:: Node * pNode=osgDB:: readNodeFile ("cow. osg") to load cow.osg, and then using the method osgEarth:: Registry: shaderGenerator(). run (pNode); My pNode node is mounted under an OSG:: Group node, and then OSG:: Group is mounted under the Earth node. The opengl version is compiled using opengl3. There is a problem with the following image
![95V2L4$LZRGRW$C01M7Z`0M](https://github.com/gwaldron/osgearth/assets/47914115/ee92b166-e170-42ef-8d2d-fafe721ce65e)
